x-pack/plugins/ml/public/application/jobs/new_job/job_from_lens/create_job.ts

```diff
@@ -0,0 +1,217 @@
+  vis: LensSavedObjectAttributes,
+  startString: string,
+  endString: string,
+  query: Query,
+  filters: Filter[],
+  dataViewClient: DataViewsContract,
+  kibanaConfig: IUiSettingsClient
+) {
+  try {
+    const { jobConfig, datafeedConfig, createdBy } = await createADJobFromLensSavedObject(
+      vis,
+      query,
+      filters,
+      dataViewClient,
+      kibanaConfig
+    );
+
+    const start = moment(startString).valueOf();
+    const end = moment(endString).valueOf();
+
+    stashJobForCloning(
+      {
+        jobConfig,
+        datafeedConfig,
+        createdBy,
+        start,
+        end,
+      } as JobCreatorType,
+      true,
+      true
+    );
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(error);
+  }
+}
+
+export async function canCreateADJob(
+  vis: LensSavedObjectAttributes,
+  query: Query | undefined,
+  filters: Filter[] | undefined,
+  dataViewClient: DataViewsContract,
+  kibanaConfig: IUiSettingsClient
+) {
+  try {
+    const jobItems = await createADJobFromLensSavedObject(
+      vis,
+      query ?? getDefaultQuery(),
+      filters ?? [],
+      dataViewClient,
+      kibanaConfig
+    );
+
+    return !!jobItems;
+  } catch (error) {
+    return false;
+  }
+}
+
+async function createADJobFromLensSavedObject(
+  vis: LensSavedObjectAttributes,
+  query: Query,
+  filters: Filter[],
+  dataViewClient: DataViewsContract,
+  kibanaConfig: IUiSettingsClient
+) {
+  const dataView = await getDataViewFromLens(vis, dataViewClient);
+  if (dataView === null) {
+    throw Error('No data views can be found in the visualization.');
+  }
+
+  const { fields, timeField, splitField } = getFields(vis.state);
+
+  if (timeField.sourceField !== dataView.timeFieldName) {
+    throw Error('Selected time field must be the default time field configured for data view.');
+  }
+
+  const jobConfig = createEmptyJob();
+  const datafeedConfig = createEmptyDatafeed(dataView.title);
+
+  const mainQuery = query?.query !== '' ? query : vis.state.query;
+
+  const { combinedQuery } = createQueries(
+    { query: mainQuery, filter: filters },
+    dataView,
+    kibanaConfig
+  );
+  datafeedConfig.query = combinedQuery;
+
+  jobConfig.analysis_config.detectors = createDetectors(fields, splitField);
+
+  jobConfig.data_description.time_field = timeField.sourceField;
+  jobConfig.analysis_config.bucket_span = '15m';
+  if (splitField) {
+    jobConfig.analysis_config.influencers = [splitField.sourceField];
+  }
+
+  const createdBy =
+    splitField || jobConfig.analysis_config.detectors.length > 1
+      ? CREATED_BY_LABEL.MULTI_METRIC
+      : CREATED_BY_LABEL.SINGLE_METRIC;
+
+  return {
+    jobConfig,
+    datafeedConfig,
+    createdBy,
+  };
+}
+
+function getFields(state: LensSavedObjectAttributes['state']) {
+  const visualization = state.visualization as { layers: XYDataLayerConfig[] };
+  const indexpattern = state.datasourceStates.indexpattern as IndexPatternPersistedState;
```

Comment:
This file is reaching deeply into arbitrary Lens visualizations. There is no fundamental problem with this, but I worry at some point new things are added to Lens which break this logic in ugly ways. Maybe we should move the getters you require into the Lens code so it's maintained as part of lens plugin so it's not your responsibility to make sure it works in all cases?

Reply:
Yes I agree, it would be far nicer not to be picking apart the visualization configuration from within ML and rather be able retrieve these field and settings from an API.